dripline/extensions/modbus/ethernet_modbus_service.py

```diff
@@ -0,0 +1,160 @@
+
+        Service.__init__(self, **kwargs)
+
+        self.ip = ip_address
+        if isinstance(indexing, int):
+            self.offset = indexing
+        elif isinstance(indexing, str):
+            if indexing.lower() == 'plc':
+                self.offset = -1
+            elif indexing.lower() == 'protocol':
+                self.offset = 0
+            else:
+                raise ValueError('Invalid indexing string argument <{}>, expect <PLC> or protocol'.format(indexing))
+        else:
+            raise TypeError('Invalid indexing type <{}>, expect string or int'.format(type(indexing)))
+
+        if wordorder.lower() == 'big':
+            self.word = pymodbus.constants.Endian.BIG
+        elif wordorder.lower() == 'little':
+            self.word = pymodbus.constants.Endian.LITTLE
+        else:
+            raise ValueError('Invalid wordorder argument <{}>, expect big or little'.format(wordorder))
+
+        if byteorder.lower() == 'big':
+            self.byte = pymodbus.constants.Endian.BIG
+        elif byteorder.lower() == 'little':
+            self.byte = pymodbus.constants.Endian.LITTLE
+        else:
+            raise ValueError('Invalid byteorder argument <{}>, expect big or little'.format(byteorder))
+
+        self.client = ModbusTcpClient(self.ip)
+        self._reconnect()
+
+    def _reconnect(self):
+        '''
+        Minimal connection method.
+        TODO: Expand to call on failed read/write, and add sophistication.
+        '''
+        if self.client.connected:
+            self.client.close()
+
+        if self.client.connect():
+            logger.debug('Connected to Alicat Device.')
+        else:
+            raise ThrowReply('resource_error_connection','Failed to Connect to Alicat Device')
+
+    def read_register(self, register, n_reg, reg_type=0x04):
+        '''
+        n_reg determines the num of registers needed to express values. More n_reg are needed for higher accuracy values.
+        reg_type: Lookup the endpoint code types that your device can access and specify in the code. 
+
+        Expand as desired according to other calls in https://pymodbus.readthedocs.io/en/latest/source/client.html#modbus-calls
+        '''
+        logger.debug('Reading {} registers starting with {}'.format(n_reg, register))
+
+        try:
+            if reg_type == 0x03:
+                result = self.client.read_holding_registers(register + self.offset, n_reg)
+            elif reg_type == 0x04:
+                result = self.client.read_input_registers(register + self.offset, n_reg)
+
+            logger.info('Device returned {}'.format(result.registers))
+
+        except Exception as e: 
+            logger.debug(f'read registers failed: {e}. Attempting reconnect.')
+            self._reconnect()
+
+        if n_reg == 1:
+            return result[0]
+        else:
+            return BinaryPayloadDecoder.fromRegisters(result.registers, wordorder=self.word, byteorder=self.byte)

```

Comment:
The BinaryPayloadDecoder is removed from current versions of pymodbus. The conversion has to happen on the per Entity and not on the Service level.